<a href="https://colab.research.google.com/github/frbarclay/Python-Data-Practicals/blob/main/worksheets/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning
---



In order to effectively analyse a dataset, often we need to prepare it first.
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g.
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted*

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [1]:
import pandas as pd
df = pd.read_excel('https://github.com/frbarclay/Python-Data-Practicals/raw/main/data/happiness_2015.xlsx',0)
df.head(0)

Empty DataFrame
Columns: [Country, Region, Happiness Rank, Happiness Score, Standard Error, Economy (GDP per Capita), Family, Health (Life Expectancy), Freedom, Trust (Government Corruption), Generosity, Dystopia Residual]
Index: []

In [2]:
sorted_table = df.sort_values('Happiness Score')
sorted_table.head()

Country                           Region  Happiness Rank  \
157     Togo               Sub-Saharan Africa             158   
156  Burundi               Sub-Saharan Africa             157   
155    Syria  Middle East and Northern Africa             156   
154    Benin               Sub-Saharan Africa             155   
153   Rwanda               Sub-Saharan Africa             154   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
157            2.839         0.06727                   0.20868  0.13995   
156            2.905         0.08658                   0.01530  0.41587   
155            3.006         0.05015                   0.66320  0.47489   
154            3.340         0.03656                   0.28665  0.35386   
153            3.465         0.03464                   0.22208  0.77370   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
157                   0.28443  0.36453                        0.10731   
156                   0.22396  0.11850                        0.10062   
155                   0.72193  0.15684                        0.18906   
154                   0.31910  0.48450                        0.08010   
153                   0.42864  0.59201                        0.55191   

     Generosity  Dystopia Residual  
157     0.16681            1.56726  
156     0.19727            1.83302  
155     0.47179            0.32858  
154     0.18260            1.63328  
153     0.22628            0.67042

In [3]:
print('The lowest score (displayed first) is',sorted_table.iloc[0,3],',',sorted_table.iloc[0,0])
print('The highest score (displayed last) is', sorted_table.iloc[157,3],',',sorted_table.iloc[157,0])

The lowest score (displayed first) is 2.839 , Togo
The highest score (displayed last) is 7.587 , Switzerland


### Exercise 2 - sort by multiple columns, display the first 5 rows
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
2. display the first 5 rows of sorted data

**Test output**:  
Records 122, 127, 147, 100, 96

In [4]:
sorted_data = df.sort_values(['Health (Life Expectancy)','Economy (GDP per Capita)'])
sorted_data.head(5)

Country              Region  Happiness Rank  \
122              Sierra Leone  Sub-Saharan Africa             123   
127                  Botswana  Sub-Saharan Africa             128   
147  Central African Republic  Sub-Saharan Africa             148   
100                 Swaziland  Sub-Saharan Africa             101   
96                    Lesotho  Sub-Saharan Africa              97   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
122            4.507         0.07068                   0.33024  0.95571   
127            4.332         0.04934                   0.99355  1.10464   
147            3.678         0.06112                   0.07850  0.00000   
100            4.867         0.08742                   0.71206  1.07284   
96             4.898         0.09438                   0.37545  1.04103   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
122                   0.00000  0.40840                        0.08786   
127                   0.04776  0.49495                        0.12474   
147                   0.06699  0.48879                        0.08289   
100                   0.07566  0.30658                        0.03060   
96                    0.07612  0.31767                        0.12504   

     Generosity  Dystopia Residual  
122     0.21488            2.51009  
127     0.10461            1.46181  
147     0.23835            2.72230  
100     0.18259            2.48676  
96      0.16388            2.79832

### Exercise 3 - sorting in descending order
---

Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [5]:
sorted_df = df.sort_values (['Freedom','Trust (Government Corruption)'],ascending=False)
sorted_df.tail(5)

Country                           Region  Happiness Rank  \
136                  Angola               Sub-Saharan Africa             137   
117                   Sudan               Sub-Saharan Africa             118   
95   Bosnia and Herzegovina       Central and Eastern Europe              96   
101                  Greece                   Western Europe             102   
111                    Iraq  Middle East and Northern Africa             112   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
136            4.033         0.04758                   0.75778  0.86040   
117            4.550         0.06740                   0.52107  1.01404   
95             4.949         0.06913                   0.83223  0.91916   
101            4.857         0.05062                   1.15406  0.92933   
111            4.677         0.05232                   0.98549  0.81889   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
136                   0.16683  0.10384                        0.07122   
117                   0.36878  0.10081                        0.14660   
95                    0.79081  0.09245                        0.00227   
101                   0.88213  0.07699                        0.01397   
111                   0.60237  0.00000                        0.13788   

     Generosity  Dystopia Residual  
136     0.12344            1.94939  
117     0.19062            2.20857  
95      0.24808            2.06367  
101     0.00000            1.80101  
111     0.17922            1.95335

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv
2. check if any NA values exist in the dataframe and print the result
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)



In [6]:
df = pd.read_csv('https://github.com/frbarclay/Python-Data-Practicals/raw/main/data/housing_in_london_yearly_variables.csv')
df.isnull().values.any()

True

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 5 - remove null values
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [8]:
cleaned_df = df.dropna(subset = ['life_satisfaction'])
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 352 entries, 613 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               352 non-null    object 
 1   area               352 non-null    object 
 2   date               352 non-null    object 
 3   median_salary      347 non-null    float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        352 non-null    object 
 6   recycling_pct      328 non-null    object 
 7   population_size    350 non-null    float64
 8   number_of_jobs     352 non-null    float64
 9   area_size          272 non-null    float64
 10  no_of_houses       272 non-null    float64
 11  borough_flag       352 non-null    int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 35.8+ KB


In [9]:
cleaned_df = df.dropna()
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 267 entries, 613 to 1013
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               267 non-null    object 
 1   area               267 non-null    object 
 2   date               267 non-null    object 
 3   median_salary      267 non-null    float64
 4   life_satisfaction  267 non-null    float64
 5   mean_salary        267 non-null    object 
 6   recycling_pct      267 non-null    object 
 7   population_size    267 non-null    float64
 8   number_of_jobs     267 non-null    float64
 9   area_size          267 non-null    float64
 10  no_of_houses       267 non-null    float64
 11  borough_flag       267 non-null    int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 27.1+ KB


## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])`

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01*

In [10]:
single_year_df = df.drop_duplicates(subset = ['area'])
print(len(single_year_df),
single_year_df['date'].unique())

51 ['1999-12-01']


# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

I started with making a function without the pandas apply function. Realising can use df.apply makes it lots easier

```
import numpy as np
def normalise_income(df):
  df['income_level'] = 0
  for i in list(range(0,len(df['median_salary']))):
    if df['median_salary'][i] < 27441:
      df['income_level'][i] = 1
    elif df['median_salary'][i] < 30932:
      df['income_level'][i] = 2
    elif df['median_salary'][i] >=30932 :
      df['income_level'][i] = 3
    else:
      df['income_level'][i] = np.nan
  return df
newdf = normalise_income(df)
newdf [['area','median_salary','income_level']]
```



In [11]:
import numpy as np
def normalise_income(df):
  if df['median_salary'] < 27441:
    return 1
  elif df['median_salary'] <30932:
    return 2
  elif df['median_salary'] >= 30932:
    return 3
  else:
    return np.nan
df['income_level'] = df.apply(normalise_income, axis = 1)
df[['area','date','median_salary','income_level']].sort_values(['area','date']).head(21)

area        date  median_salary  income_level
1     barking and dagenham  1999-12-01        21480.0           1.0
52    barking and dagenham  2000-12-01        22618.0           1.0
103   barking and dagenham  2001-12-01        22323.0           1.0
154   barking and dagenham  2002-12-01        24813.0           1.0
205   barking and dagenham  2003-12-01        25358.0           1.0
256   barking and dagenham  2004-12-01        26089.0           1.0
307   barking and dagenham  2005-12-01        26680.0           1.0
358   barking and dagenham  2006-12-01        26549.0           1.0
409   barking and dagenham  2007-12-01        30200.0           2.0
460   barking and dagenham  2008-12-01        29396.0           2.0
511   barking and dagenham  2009-12-01        30973.0           3.0
562   barking and dagenham  2010-12-01        28527.0           2.0
613   barking and dagenham  2011-12-01        28201.0           2.0
664   barking and dagenham  2012-12-01        33131.0           3.0
715   barking and dagenham  2013-12-01        32248.0           3.0
766   barking and dagenham  2014-12-01        32698.0           3.0
817   barking and dagenham  2015-12-01        33018.0           3.0
868   barking and dagenham  2016-12-01        30994.0           3.0
919   barking and dagenham  2017-12-01        28553.0           2.0
970   barking and dagenham  2018-12-01        29995.0           2.0
1021  barking and dagenham  2019-12-01        28738.0           2.0

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [16]:
median =  df.describe()['number_of_jobs']['50%']
UQ =  df.describe()['number_of_jobs']['75%']

In [17]:
def normalise_jobs(df):
  if df['number_of_jobs'] < median:
    return 1
  elif df ['number_of_jobs'] < UQ:
    return 2
  elif df ['number_of_jobs'] >= UQ:
    return 3
  else:
    return np.nan
df['jobs_level']= df.apply(normalise_jobs, axis =1)

In [18]:
df[['area','date','number_of_jobs','jobs_level']].dropna()

area        date  number_of_jobs  jobs_level
51          city of london  2000-12-01        361000.0         2.0
52    barking and dagenham  2000-12-01         57000.0         1.0
53                  barnet  2000-12-01        138000.0         1.0
54                  bexley  2000-12-01         76000.0         1.0
55                   brent  2000-12-01        122000.0         1.0
...                    ...         ...             ...         ...
1015         great britain  2018-12-01      34850000.0         3.0
1016     england and wales  2018-12-01      31989000.0         3.0
1017      northern ireland  2018-12-01        900000.0         2.0
1018              scotland  2018-12-01       2861000.0         3.0
1019                 wales  2018-12-01       1496000.0         2.0

[931 rows x 4 columns]

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1

In [ ]:
median =  df.describe()['number_of_jobs']['50%']
UQ =  df.describe()['number_of_jobs']['75%']

In [20]:
def normalise_housing(df):
  if df['no_of_houses'] < median:
    return 1
  elif df ['no_of_houses'] < UQ:
    return 2
  elif df ['no_of_houses'] >= UQ:
    return 3
  else:
    return np.nan
df['housing_volume']= df.apply(normalise_jobs, axis =1)
df[['area','date','no_of_houses','housing_volume']].dropna()

area        date  no_of_houses  housing_volume
102         city of london  2001-12-01        5009.0             2.0
103   barking and dagenham  2001-12-01       68298.0             1.0
104                 barnet  2001-12-01      130515.0             1.0
105                 bexley  2001-12-01       91606.0             1.0
106                  brent  2001-12-01      101427.0             1.0
...                    ...         ...           ...             ...
999         waltham forest  2018-12-01      103029.0             1.0
1000            wandsworth  2018-12-01      146162.0             1.0
1001           westminster  2018-12-01      124509.0             2.0
1008                london  2018-12-01     3556161.0             3.0
1013               england  2018-12-01    24172166.0             3.0

[630 rows x 4 columns]

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [47]:
mean_area = df.dropna().describe()['area_size']['mean']

In [48]:
def normalise_area(df):
  if df['area_size'] < mean_area:
    return 0
  elif df['area_size'] >= mean_area:
    return 1
  else:
    return np.nan
df['area_size_normalised'] = df.dropna().apply(normalise_area, axis =1)
df[['area','area_size_normalised']].dropna().head(35)

area  area_size_normalised
613    barking and dagenham                   0.0
614                  barnet                   0.0
615                  bexley                   0.0
616                   brent                   0.0
617                 bromley                   0.0
618                  camden                   0.0
619                 croydon                   0.0
620                  ealing                   0.0
621                 enfield                   0.0
622               greenwich                   0.0
623                 hackney                   0.0
624  hammersmith and fulham                   0.0
625                haringey                   0.0
626                  harrow                   0.0
627                havering                   0.0
628              hillingdon                   0.0
629                hounslow                   0.0
630               islington                   0.0
631  kensington and chelsea                   0.0
632    kingston upon thames                   0.0
633                 lambeth                   0.0
634                lewisham                   0.0
635                  merton                   0.0
636                  newham                   0.0
637               redbridge                   0.0
638    richmond upon thames                   0.0
639               southwark                   0.0
640                  sutton                   0.0
641           tower hamlets                   0.0
642          waltham forest                   0.0
643              wandsworth                   0.0
644             westminster                   0.0
651                  london                   0.0
656                 england                   1.0
664    barking and dagenham                   0.0